lib/src/repository/git_local_repository.dart

```diff
@@ -0,0 +1,159 @@
+// Copyright (c) 2022, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:io';
+
+import 'package:retry/retry.dart';
+
+import '../logging.dart';
+import '../utils.dart' show runProc, withTempDir, PanaProcessResult;
+
+final _acceptedBranchNameRegExp = RegExp(r'^[a-z0-9]+$');
+
+/// Detects the name of the default git branch on the given [baseUrl],
+/// using the metadata returned by `git remote show`.
+///
+/// Returns `null` when no branch name can be detected, or the implied
+/// branch name does not conform of the allowed patterns.
+Future<String?> tryDetectDefaultGitBranch(String baseUrl) async {
+  return await withTempDir((dir) async {
+    GitLocalRepository? repo;
+    try {
+      repo = await GitLocalRepository.createLocalRepository(baseUrl);
+      return await repo.detectDefaultBranch();
+    } on GitToolException catch (e, st) {
+      log.warning(e.toString(), st);
+      return null;
+    } finally {
+      await repo?.delete();
+    }
+  });
+}
+
+class GitLocalRepository {
```

Comment:
What is this, can we give a documentation comment? should we annotated it `@sealed`?

```suggestion
/// Interface for reading a remote git repository.
///
/// This objects uses a local temporary folder for interfacing with remote repository.
/// Hence, it is important to call [delete] or temporary files will be leaked.
class GitLocalRepository {
```